R/layers.R

```diff
@@ -102,40 +140,240 @@ sass_layer <- function(defaults = "", declarations = "", rules = "",
+# @param name Sass layer name to use inside the Sass bundle object
+as_sass_bundle <- function(x, name = "") {
+  stopifnot(is.character(name) && length(name) == 1)
+
+  # Upgrade pattern:
+  # is_sass_bundle(x) && name == "" -> x
+  # is_sass_bundle(x) && name != "" -> sass_bundle(!!name := as_sass_layer(x))
+  # is_sass_layer(x)                -> sass_bundle(!!name := x)
+  # is.list(x) && has_any_name(x)   -> sass_bundle(!!name := sass_layer(defaults = x))
+  # else                            -> sass_bundle(!!name := sass_layer(rules = x))
+
+  # A list of sass_layer values can be handled by sass_bundle(...) or sass_bundle(!!!x). Do not test for this
+
+  if (is_sass_bundle(x)) {
+    # if there is nothing special about the name, return
+    if (identical(name, "")) {
+      return(x)
+    }
+
+    # convert to a single sass layer object so the overall name can be used
+    x <- as_sass_layer(x)
+  }
+
+  single_layer <-
+    if (is_sass_layer(x)) {
+      x
+    } else {
+      # upgrade via sass_layer
+      if (is.list(x) && has_any_name_recursive(x)) {
+        # a name was found somewhere
+        sass_layer_struct(defaults = x)
+      } else {
+        sass_layer_struct(rules = x)
+      }
+    }
+
+  layers <- list()
+  layers[[name]] <- single_layer
+
+  sass_bundle_struct(layers = layers)
+}
+
+#' Helps avoid sass_bundle inf recursion
+#' @param layers named or unnamed list containing sass_layer objects
+#' @return object of class `sass_bundle`
+#' @noRd
+sass_bundle_struct <- function(layers = list()) {
+  layers_are_sass_layers <- vapply(layers, is_sass_layer, logical(1))
+  stopifnot(all(layers_are_sass_layers))
+  ret <- list(
+    layers = layers
+  )
+  class(ret) <- "sass_bundle"
+  ret
+}
+
+#' @describeIn sass_layer Collect `sass_bundle()` and/or `sass_layer()` objects. Unnamed Sass bundles will be concatenated together, preserving their internal name structures. Named Sass bundles will be condensed into a single Sass layer for easier removal from the returned Sass bundle.
+#' @export
+sass_bundle <- function(...) {
+  layers <- dropNulls(list2(...))
+  layers_upgraded <-
+    mapply(
+      SIMPLIFY = FALSE,
+      layers,
+      names2(layers),
+      FUN = function(x, name) {
+        as_sass_bundle(x, name = name)
+      }
+    )
+  # collect and flatten
+  ## unlist(list(list(1), list(), NULL, list(2)), recursive = FALSE)
+  #> [[1]]
+  #> [1] 1
+  #>
+  #> [[2]]
+  #> [1] 2
+  ret_layers <- unlist(
+    lapply(unname(layers_upgraded), `[[`, "layers"),
+    recursive = FALSE
+  )
+  sass_bundle_struct(ret_layers)
+}
+
+
+#' @describeIn sass_layer Remove a whole `sass_layer()` from a `sass_bundle()` object.
+#' @param bundle Output value from `sass_layer()` or `sass_bundle()`
+#' @param name If a Sass layer name is contained in `name`, the matching Sass layer will be removed from the `bundle`
+#' @export
+sass_bundle_remove <- function(bundle, name) {
+  stopifnot(is_sass_bundle(bundle))
```

Comment:
Seems we should throw here as well if `identical(name, "")`

Reply:
Yes! Good catch!